abn/watcher/app.go

```diff
@@ -0,0 +1,189 @@
+package watcher
+
+// app.go - methods to track the runtime state of applications and their versions
+// For each version, maintain information about its readiness and its mapping to a "track", if any.
+
+// A track is a (user assigned) identifier that the user assigns to versions as part of the CI/CD process.
+// When Iter8 A/N(/n) service is used to lookup versions, the track identifier is returned.
+// The caller can use this to route requests to the appropriate version.
+// To do this, the set of track identifiers should be a (small) fixed set, such as "current" and
+// "candidate", that can be mapped to a static routes.
+
+import (
+	"github.com/iter8-tools/iter8/abn/appsummary"
+	"github.com/iter8-tools/iter8/base/log"
+)
+
+// Application is runtime information about the versions of an application
+type Application struct {
```

Comment:
There seem to be two in-memory data structures for caching app data. One in watcher, and another in appsummary. Could we please coalesce them into one? We could add additional methods to the latter if need be.

Reply:
I considered this. However, the summary object is one that can we directly converted to/from content of secret. The details in the runtime object are not included there. Do you know if skipping some `json` and `yaml` tags will make this work? I hadn't thought of it before.

Reply:
What are the additional information that the `runtime object` contains which the summary object does not contain? Can those details be constructed simply using additional methods on `ApplicationData` (instead of additional fields)? Clearly, there is a need to keep the two structures consistent, which means there's already some need to go back and forth (in terms of functions/methods) between these two data structures.

Reply:
This does not work.

Reply:
I am thinking more about this and am inclined to think the objects while sharing name and both expressing atttrbiutes of an application (a colletion of versions), they are used quite distinctly. In one use case they are an in memory consturct meant to record observed versions in the cluster. In the second, they are a persistent record of the metrics associated with each version and a history. Here they are read/modified/written (to update) or just read (for assessment). I do not see this as a good fit to merge into a single object. However, I wonder if there is naming that might make their purpose clearer. I tried with Application and ApplicationSummary but the child version objects have the same name.

Reply:
1. record observed versions in the cluster
2. record of [...] and a history.

Isn't 2 a superset of 1? Can't 1 be constructed from the recent history? Doesn't storing this information twice in-memory twice mean that additional code needs to be written to keep them consistent?

Reply:
Further, when we add the task, I expect the contents of current appsummary to move to base (base needs a notion of summarymetric and the mechanism to read it. I would not expect any of the aspects of the runtime application object ot be present in base.

Reply:
Also, everything we're discussing so far is about caching for reads. Ultimately, we need to think about write caches also (for metrics), since this is required for scalability.

Reply:
> I would not expect any of the aspects of the runtime application object to be present in base.

Correct! 

The idiomatic way to achieve this in Golang is as follows.

```go
package base

type AppSummary struct {...}

func ReadAppSummary(c ClientSet, appName string) AppSummary
```

```go
package abn

type AppSummary base.AppSummary

// more funcs and methods using abn.AppSummary; you still can base.ReadAppSummary here ... to achieve code reuse
```

Reply:
The runtime and persistent forms of application have been merged together in `abn/application`. A clear part of the model is the part persisted.